File: 2021/day-02/2021-02-2.py
#!/usr/bin/env python3
import sys
sys.path.append("../../libs/python")
import n0s1_aoc as n0s1

def get_pos(data):
    depth = 0
    horizontal = 0
    aim = 0
    for line in data[:-1].split("\n"):
        d = line.split(" ")
        if d[0] == "forward":
            horizontal += int(d[1])
            depth += (int(d[1]) * aim)
        elif d[0] == "down":
            aim += int(d[1])
        elif d[0] == "up":
            aim -= int(d[1])
    return depth * horizontal

print(get_pos(n0s1.file_to_string("sample-2021-02.txt")))
print(get_pos(n0s1.file_to_string("input-2021-02.txt")))

File: 2021/day-02/2021-02-2.ex
defmodule Day2 do

  def get_position(fname) do
    case File.read(fname) do
      {:ok, body} -> String.trim(body)
      |> String.split("\n")
      |> Enum.map(fn x -> [k | [v | _]] = String.split(x)
          [String.to_atom(k), String.to_integer(v)] end)
      |> calc_pos
    {:error, reason} -> IO.puts "Error reading file: #{reason}"
    end
  end

  def calc_pos(ht, x \\ 0, y \\ 0, aim \\ 0)
  def calc_pos([], x, y, _), do: x * y
  def calc_pos([[d | [v | _]] | t], x, y, aim) do
    case d do
      :forward -> calc_pos(t, x + v, y + (v * aim), aim)
      :down ->    calc_pos(t, x, y, aim + v)
      :up ->      calc_pos(t, x, y, aim - v)
    end
  end

  def main do
    IO.puts get_position("sample-2021-02.txt")
    IO.puts get_position("input-2021-02.txt")
  end

end
